### DISCUSSION TITLE
Whether there would be a performance issue related to large array returns from Smart Contract getter functions?

### DISCUSSION THREAD

**Discussion by ChanX21** [Cyfrin/foundry-full-course-cu#2588] [General]
Hey guys, just a quick question. Suppose my getter function in my contract returns an array of say 10-20 million values. 

a) Would it cause performance issues while I call it from my Node.Js backend or React frontend?
b) What would be the alternate solution if there is a performative issue?

**Comment by EngrPips** [comment]
Hello @ChanX21, If your application require having an array that will hold 10-20 million value maybe it is better to use a mapping on the smart contract side and keep the array to your front-end. that way you won't need to get that much large data from a smart contract. But maybe there is some more clever work around. Let's look forward to people suggestions.

**Reply by ChanX21** [reply]
haha yeah that is what I'm waiting for. Also regarding the reason why we are using the getter that way is because we have to get from the SC all the user who have say minted an NFT and there is no way we could get that from using mapping unless of course if we give the key as numbers from 0 to n but then we just replicated arrays

**Reply by EngrPips** [reply]
First, you should understand that you don't need to store all your application data on your smart contract. A smart contract is not a database where you store anything and everything. You only need to store the data the smart contract requires to run properly, So any data that is primarily needed by your front end should be stored on the front end except if it is needed by the smart contract. If you want to get every user that has minted an NFT so as to maybe display some stuff on the front end, then you can keep that array to the front end, but if you need that list for some computation in your smart contract, then you might want to consider redesigning your smart contract as an array can grow too big that looping over them for an operation will cause out of gas exception.

**Reply by ChanX21** [reply]
Yeah understood your point @EngrPips, yeah agree SC's are not ideally meant for large data storage but again what is stopping it is my question. Like say the gas cost to read 1 data or 1 billion data from the SC is the same ( ie 0 ). I know something is off in my approach but i can't figure it out. 

Also the alternative to this approach could be indexing from the backend and reading the data of the DB but again since the SC reader functions are basically zero gas costing why bother with it is the Q.

**Reply by EngrPips** [reply]
Yeah, I understand perfectly your point, and I must say it is reasonable to ask why bother when reading from the blockchain is free but if it is slowing down your process maybe try with a database and see if there will be any improvement.
